backend/src/statistics/statistics.service.ts

```diff
@@ -0,0 +1,207 @@
+    const data = await this.prisma.roomSession.findMany({
+      where: { users: { some: { id: userId } }, endTime: { not: null } },
+      include: {
+        question: {
+          select: { title: true },
+        },
+        users: true,
+      },
+    });
+    return data.map((session) => {
+      return {
+        userName: session.users.find((user) => user.id !== userId)?.name || "",
+        questionTitle: session.question.title,
+        date: session.startTime,
+      };
+    });
+  }
+
+  private async getHeatmapDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+        startTime: {
+          gte: subYears(new Date(), 1),
+        },
+      },
+    });
+    return data.map((session) => {
+      return { date: session.startTime };
+    });
+  }
+
+  private async getNetworkDataByUserId(userId: number) {
+    const data = await this.prisma.roomSession.findMany({
+      where: {
+        users: { some: { id: userId } },
+        endTime: { not: null },
+      },
+      include: {
+        question: {
+          include: {
+            topics: true,
+          },
+        },
+      },
+    });
+
+    const existing = new Set<number>();
+    const topics = new Map<number, string>();
+    const relations = new Map<number, Map<number, number>>();
+    for (const session of data) {
+      // We only want to count unique questions.
+      if (existing.has(session.questionId)) {
+        continue;
+      }
+      existing.add(session.questionId);
+      const sortedTopics = session.question.topics.sort(
+        (lhs, rhs) => lhs.id - rhs.id,
+      );
+      for (const topic of sortedTopics) {
+        topics.set(topic.id, topic.name);
+      }
+
+      for (let low = 0; low < sortedTopics.length; low++) {
+        for (let high = low + 1; high < sortedTopics.length; high++) {
+          const currentMap = relations.get(sortedTopics[low].id);
+          const innerMap = currentMap || new Map<number, number>();
+          innerMap.set(
+            sortedTopics[high].id,
+            (innerMap.get(sortedTopics[high].id) || 0) + 1,
+          );
+          if (!currentMap) {
+            relations.set(sortedTopics[low].id, innerMap);
+          }
+        }
+      }
+    }
+
+    const topicSize = new Map<number, number>();
+    for (const outerEntry of relations) {
+      const lowId = outerEntry[0];
+      for (const innerEntry of outerEntry[1]) {
+        const highId = innerEntry[1];
+        topicSize.set(lowId, (topicSize.get(lowId) || 0) + 1);
+        topicSize.set(highId, (topicSize.get(highId) || 0) + 1);
+      }
+    }
+
+    return {
+      topics: Array.from(topics.entries()).map((entry) => ({
+        id: entry[0],
+        name: entry[1],
+        count: topicSize.get(entry[0]) || 0,
+      })),
+      relations: Array.from(relations.entries()).flatMap((edges) => {
```

Comment:
nit: can we call this "links" instead of "relations"? So that I do not have to perform excessive processing on my end apart from changing count to symbolSize and value